sql: rationalize the handling of comments and improve performance

rafiss opened the pull request:
rebase of https://github.com/cockroachdb/cockroach/pull/95431
fixes https://github.com/cockroachdb/cockroach/issues/102851

Prior to this patch, we had a weird translation going in
`kv_catalog_comments`, which was causing the tree of objects to be
traversed 6 times (one per comment type), and then the rows to be
mapped one by one again in `pg_description` / `pg_shdescription`.

This made no sense - it should be possible to push a predicate on
`pg_description` down to a virtual index on `kv_catalog_comments`, and
this requires the column types to align and no computation to happen
in-between.

This commit simplifies all this by having `kv_catalog_comments`
responsible for preparing the data in the format expected by the `pg_`
tables.

The number of roundtrips needed for an ORM query that uses
pg_description dropped to 4, from 134.

```
goos: linux
goarch: amd64
cpu: Intel(R) Xeon(R) CPU @ 2.80GHz
                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │                sec/op                 │                   sec/op                    vs base                │
ORMQueries/introspection_description_join-24                           10687.1m ± 1%                                  144.1m ± 8%  -98.65% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │              roundtrips               │                 roundtrips                  vs base                │
ORMQueries/introspection_description_join-24                            134.000 ± 0%                                   4.000 ± 0%  -97.01% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │                 B/op                  │                    B/op                     vs base                │
ORMQueries/introspection_description_join-24                          3846.69Mi ± 0%                                 38.98Mi ± 6%  -98.99% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │               allocs/op               │                 allocs/op                   vs base                │
ORMQueries/introspection_description_join-24                           45566.9k ± 0%                                  433.3k ± 6%  -99.05% (p=0.000 n=10)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104763)
<!-- Reviewable:end -->

rafiss committed:
sql: rationalize the handling of comments

Prior to this patch, we had a weird translation going in
`kv_catalog_comments`, which was causing the tree of objects to be
traversed 6 times (one per comment type), and then the rows to be
mapped one by one again in `pg_description` / `pg_shdescription`.

This made no sense - it should be possible to push a predicate on
`pg_description` down to a virtual index on `kv_catalog_comments`, and
this requires the column types to align and no computation to happen
in-between.

This commit simplifies all this by having `kv_catalog_comments`
responsible for preparing the data in the format expected by the `pg_`
tables. It also adds kv_builtin_function_comments, so that the large
number of predefined comments on builtin functions don't cause joins on
the table to be very large in the common case.

The number of roundtrips needed for an ORM query that uses
pg_description dropped to 4, from 134.

```
goos: linux
goarch: amd64
cpu: Intel(R) Xeon(R) CPU @ 2.80GHz
                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │                sec/op                 │                   sec/op                    vs base                │
ORMQueries/introspection_description_join-24                           10687.1m ± 1%                                  144.1m ± 8%  -98.65% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │              roundtrips               │                 roundtrips                  vs base                │
ORMQueries/introspection_description_join-24                            134.000 ± 0%                                   4.000 ± 0%  -97.01% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │                 B/op                  │                    B/op                     vs base                │
ORMQueries/introspection_description_join-24                          3846.69Mi ± 0%                                 38.98Mi ± 6%  -98.99% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │               allocs/op               │                 allocs/op                   vs base                │
ORMQueries/introspection_description_join-24                           45566.9k ± 0%                                  433.3k ± 6%  -99.05% (p=0.000 n=10)
```

Release note: None

postamar commented on pkg/sql/logictest/testdata/logic_test/information_schema:
```diff
@@ -2949,20 +2952,20 @@ SELECT table_catalog, table_schema, table_name, column_name, column_comment
+test           public        t1          id           identification
+test           public        t1          name         character
+test           public        t1          class        NULL
+test           public        t1          rowid        NULL
```
Do we want NULLs here, or should those be replaced by empty strings? I honestly can't remember exactly. Feel free to disregard if this doesn't matter.

postamar reviewed (commented):
Very nice!

rafiss commented on pkg/sql/crdb_internal.go:
```diff
@@ -5443,41 +5475,56 @@ CREATE TABLE crdb_internal.kv_catalog_comments (
+					return err
+				}
+				classOid = tree.NewDOid(catconstants.PgCatalogConstraintTableID)
+				objOid = getOIDFromConstraint(c, db.GetID(), schema.GetID(), tableDesc)
```
i had just copied the logic from `pkg/sql/pg_catalog.go` so i can't take credit for the nice idea :)

the issue you're describing is pre-existing, but i can fix it

cucaroach reviewed (commented):
Seems fine from the bits I understood.   Nice speedup!

> Reviewed 7 of 21 files at r1, 10 of 15 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104763#-:-NXyyiNy2w1zpkaLnLM3:b-amblnl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @postamar, and @rafiss)

___
*[`-- commits` line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104763#-NXytEvJ-pmy2T1jyZRz):*
Not sure what happened here but looks like some whitespace got eaten or something.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104763)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @fqazi, and @postamar)

___
*[`-- commits` line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104763#-NXytEvJ-pmy2T1jyZRz):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Not sure what happened here but looks like some whitespace got eaten or something.
</blockquote></details>

![image.png](https://files.reviewableusercontent.io/images/reviewable/1320573/_FvAs9QodO-5xMagzQrbAttp/image.png)

hm, it seems fine to me. you might need to make your browser window wider to view the table easily.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
@postamar this is ready for another look

rafiss commented:
tftr!

bors r+
